proposals/mainnet_migration.md

```diff
@@ -0,0 +1,528 @@
+
+The migration from Lisk Core v3 to Lisk Core v4, also depicted in Figure 1, is performed by the following steps:
+
+1. Nodes run Lisk Core v3, where the following steps are done:
+    1. After a block at height `HEIGHT_SNAPSHOT` was processed, a snapshot of the state is performed, which we denote by `STATE_SNAPSHOT`. If this block is reverted and a new block for this height is processed, then also `STATE_SNAPSHOT` needs to be computed again.
+    2. Nodes continue processing blocks until the block at height `HEIGHT_SNAPSHOT` is final.
+    3. Once the block at height `HEIGHT_SNAPSHOT` is final, nodes can stop forging and processing new blocks. Finalized blocks with a height larger than or equal to `HEIGHT_SNAPSHOT+1` are discarded.
+2. Nodes compute a snapshot block as defined [below](#snapshot-block) using a _migrator tool_ and persist this one locally.
+3. Nodes start to use Lisk Core v4, where the steps described in the section [Starting Lisk Core v4](#starting-lisk-core-4) are executed.
+4. Once the timeslot of the snapshot block at height `HEIGHT_SNAPSHOT+1` is passed, the first round following the new protocol starts.
+
+##### Starting Lisk Core v4
+
+When Lisk Core v4 is started for the first time, the following steps are performed:
+
+1. Get the snapshot block (the one for height `HEIGHT_SNAPSHOT+1`):
+    1. Check if the snapshot block for height `HEIGHT_SNAPSHOT+1` exists locally. If yes, fetch this block. If not, stop here.
+2. Process the snapshot block as described in [LIP 0060](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0060.md).
+3. Check if a database from Lisk Core v3 containing all blocks between heights `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` and `HEIGHT_SNAPSHOT` (inclusive) can be found locally. If yes:
+    1. Fetch these blocks from highest height to lowest. Each block is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block and its transactions are persisted in the database.
+    2. Skip steps 4 and 5.
+4. Fetch all blocks between heights `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK+1` and `HEIGHT_SNAPSHOT` (inclusive) via peer-to-peer from highest height to lowest. Each block is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block and its transactions are persisted in the database.
+5. The snapshot block for the height `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` is downloaded from a server. The URL for the source can be configured in Lisk Core v4+. When downloaded, it is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block is persisted in the database.
```

Comment:
```suggestion
5. The snapshot block for the height `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` is downloaded from a server. The URL for the source can be configured. When downloaded, it is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block is persisted in the database.
```

Reply:
I don't understand why the part " in Lisk Core v4+" is necessary or correct here as it is talking about Lisk Core 4.

Reply:
The snapshot block can be downloaded from the server only after it is finalized. So here Lisk Core v4+ means a later version of Lisk Core v4, in which a link to the snapshot block is hardcoded.

This is discussed in details a bit later, in the section Fetching Snapshot Blocks, I have added a reference for clarity.

Reply:
But here we talk about the previous snapshot block (not that one at `HEIGHT_SNAPSHOT+1`, but the one at `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK`). This block is from August last year and already final. In particular, it would be safe to obtain it from an untrusted source due to the minimal validations defined below.

Reply:
Ah, you are right! Changed as per your suggestion.